### PR TITLE
Refactor directory access into DirectoryTree - FS [2/3]

### PR DESCRIPTION
**Pull request by Ghabry**
This is a huge refactor of the FileFinder.

To sum up:
- All general file access related stuff moved from FileFinder to ``DirectoryTree``. The game directory tree and specific functions such as ``FindImage`` are still managed by ``FileFinder``. Will stay this way until the next PR.
- The RTP redirection was moved to ``FileFinder_RTP``. Converted to a class and managed by ``Main_Data`` as this is game specific.
- StringView-ifyd stuff

It will not conclude my refactor though, next step after this are the Virtual File Systems

--------------

The ``DirectoryTree`` uses now unique instead of shared_ptr. With ``Subtree`` non-owning subviews of the tree (``DirectoryTreeView``) can be created.
E.g. the Game Browser can just pass a Subtreeview to the game it boots, this way there is no need anymore for enumerating twice on startup -> faster startup times on slow systems.
Though the ownership semantics are quite messy in Player so this is not possible yet and I postbone it to the next PR.

These subviews will be also useful for the Virtual Filesystems later.

--------------

TODO:

Retest:

- [x] Fix Windows
- [x] Game Browser Game booting
- [ ] Meta File (Game Import)
- [x] Extensions guessing
- [x] MP3 major game heuristic
- [x] Random game testing (Frozen Trigger works)
- [x] Translation support 
- [x] File not found debug log is now very common because there is only one find entry point that handles everything-
- [x] Fix ownership of RTP trees.
- [x] Documentation
- [x] Testing

### CODE REVIEW

**Comment by Ghabry** [comment]
Surprisingly emscripten still works

**Comment by Ghabry** [comment]
just a rebase. Translation support disabled for now while I think about a nice way to hook it into my API.

**Comment by Ghabry** [comment]
This is now ready for review :).

Only things I want to do is write some doc comments and fix two pending problems:

- I can't get the "Save Import from previous game" (Scene Import) Importer to work
- Windows path handling (the backslash) is broken according to a unit test

**Comment by Ghabry** [comment]
Obviously Windows was a one-line fix...

**Comment by Ghabry** [comment]
If you don't mind I would like to postbone the "Scene Import" Fix for "FS [3/3]".
When I fix this now I very likely have to redo the fix for the VFS, so thats extra work.

The problem is:
The Scene requests a directory tree for the folder "..".

This has multiple problems:

- The path is non-canonical
- The directory tree of the game cannot access its parent folder because that is above the game root
- Have to invent an API in the VFS to request the parent VFS

File: tests/rtp.cpp [hunk 1]
```diff
@@ -28,6 +39,7 @@ static std::shared_ptr<FileFinder::DirectoryTree> make_tree() {
 			{"カーソル1.wav", "カーソル1.wav"},
 			{"キャンセル1.wav", "キャンセル2.wav"}
 	};
+#endif
```
Comment (Ghabry) [inline tests/rtp.cpp:42]: oh forgot to fix these

File: tests/filefinder.cpp [hunk 2]
```diff
@@ -28,19 +28,62 @@ TEST_CASE("IsDirectory") {
...
+    CHECK(file == "file");
+
+    std::tie(path, file) = FileFinder::GetPathAndFilename("file");
+    CHECK(path == "");
+    CHECK(file == "file");
+
+    std::tie(path, file) = FileFinder::GetPathAndFilename("folder/folder2/file");
+    CHECK(path == "folder/folder2");
+    CHECK(file == "file");
+}
```
Comment (Ghabry) [inline tests/filefinder.cpp:87]: whitespace instead of tabs...

File: src/directory_tree.cpp [hunk 3]
```diff
@@ -0,0 +1,215 @@
...
+		const std::string name = dir.GetEntryName();
+		Platform::FileType type = dir.GetEntryType();
+
+		static bool has_fast_dir_stat = true;
+		bool is_directory = false;
+		if (has_fast_dir_stat) {
+			if (type == Platform::FileType::Unknown) {
+				has_fast_dir_stat = false;
+			} else {
+				is_directory = type == Platform::FileType::Directory;
```
Comment (fmatthew5876) [inline src/directory_tree.cpp:123]: Put parens around the expression

File: src/directory_tree.cpp [hunk 4]
```diff
@@ -0,0 +1,291 @@
...
+	if (!dir) {
+		Output::Debug("Error opening dir {}: {}", full_path, ::strerror(errno));
+		return nullptr;
+	}
+
+	while (dir.Read()) {
+		const std::string name = dir.GetEntryName();
+		Platform::FileType type = dir.GetEntryType();
+
+		static bool has_fast_dir_stat = true;
```
Comment (fmatthew5876) [inline src/directory_tree.cpp:117]: Whats the reason for this bool?

Comment (fmatthew5876) [inline src/directory_tree.cpp:117]: This dance with the static bool is confusing, is it really necessary? Why not just say 
```
if (type == Directory) {
  is_directory = true;
} else if (type == Unknown) {
  is_directory = FileFinder::IsDirectory(...);
}
```

File: src/directory_tree.cpp [hunk 5]
```diff
@@ -0,0 +1,291 @@
...
+			} else {
+				is_directory = type == Platform::FileType::Directory;
+			}
+		}
+
+		if (!has_fast_dir_stat) {
+			is_directory = FileFinder::IsDirectory(FileFinder::MakePath(full_path, name), true);
+		}
+
+		if (name == "." || name == "..") {
```
Comment (fmatthew5876) [inline src/directory_tree.cpp:131]: Move this check up so you're not doing filesystem calls on `.` and `..`

File: src/directory_tree.cpp [hunk 6]
```diff
@@ -0,0 +1,291 @@
...
+	}
+
+	Platform::Directory dir(full_path);
+	if (!dir) {
+		Output::Debug("Error opening dir {}: {}", full_path, ::strerror(errno));
+		return nullptr;
+	}
+
+	while (dir.Read()) {
+		const std::string name = dir.GetEntryName();
```
Comment (fmatthew5876) [inline src/directory_tree.cpp:114]: `const auto&` ?

File: src/directory_tree.cpp [hunk 7]
```diff
@@ -0,0 +1,291 @@
...
+	dir_cache[dir_key] = fs_path;
+
+	DirectoryListType fs_cache_entry;
+
+	for (auto& entry : entries) {
+		fs_cache_entry.emplace(std::make_pair(make_key(entry.name), entry));
+	}
+	fs_cache.emplace(dir_key, fs_cache_entry);
+
+	dir_it = dir_cache.find(dir_key);
```
Comment (fmatthew5876) [inline src/directory_tree.cpp:159]: `dir_it` is unused?

File: src/directory_tree.cpp [hunk 8]
```diff
@@ -0,0 +1,291 @@
...
+
+DirectoryTreeView DirectoryTree::Subtree(std::string sub_path) {
+	return DirectoryTreeView(this, std::move(sub_path));
+}
+
+std::string DirectoryTree::FindFile(StringView filename, Span<StringView> exts) const {
+	return FindFile({ ToString(filename), exts });
+}
+
+std::string DirectoryTree::FindFile(StringView directory, StringView filename, lcf::Span<StringView> exts) const {
```
Comment (fmatthew5876) [inline src/directory_tree.cpp:172]: `Span` instead of `lcf::Span`

File: src/directory_tree.cpp [hunk 9]
```diff
@@ -0,0 +1,291 @@
...
+	if (!entries) {
+		if (args.file_not_found_warning) {
+			Output::Debug("Cannot find: {}/{}", dir, name);
+		}
+		return "";
+	}
+
+	std::string dir_key = make_key(dir);
+	auto dir_it = dir_cache.find(dir_key);
+	assert(dir_it != dir_cache.end());
```
Comment (fmatthew5876) [inline src/directory_tree.cpp:210]: `dir_it` not used except for in the assert, I would wrap it all in `NDEBUG` so the `dir_cache` lookup isn't done in release build

File: src/directory_tree.cpp [hunk 10]
```diff
@@ -0,0 +1,291 @@
...
+	return FileFinder::MakePath(root, subpath);
+}
+
+StringView DirectoryTree::GetRootPath() const {
+	return root;
+}
+
+DirectoryTreeView::DirectoryTreeView(const DirectoryTree* tree, std::string sub_path) :
+	tree(tree), sub_path(std::move(sub_path)) {
+	valid = tree->ListDirectory(this->sub_path) != nullptr;
```
Comment (fmatthew5876) [inline src/directory_tree.cpp:253]: parens

File: src/directory_tree.h [hunk 11]
```diff
@@ -0,0 +1,266 @@
...
+
+		Entry(std::string name, FileType type) : name(std::move(name)), type(type) {}
+	};
+
+	/** Argument struct for more complex find operations */
+	struct Args {
+		/** File relative to the current tree to search */
+		std::string path;
+		/** File extensions to append to the filename when searching */
+		lcf::Span<StringView> exts;
```
Comment (fmatthew5876) [inline src/directory_tree.h:63]: `Span`

File: src/directory_tree.h [hunk 12]
```diff
@@ -0,0 +1,266 @@
...
+	const DirectoryTree* tree = nullptr;
+	std::string sub_path;
+	bool valid = false;
+};
+
+inline DirectoryTreeView::operator bool() const noexcept {
+	return valid;
+}
+
+inline bool operator<(const DirectoryTree::Entry& l, const DirectoryTree::Entry& r) {
```
Comment (fmatthew5876) [inline src/directory_tree.h:250]: It's annoying, but we should provide the other operators. They can all be written in terms of `operator<` so you don't need to duplicate code.

File: src/fileext_guesser.cpp [hunk 13]
```diff
@@ -31,36 +31,43 @@ namespace {
...
-			if (ext != "exe" && ext != "ini") {
-				candidates.push_back({item.second, ext, FileFinder::GetFileSize(FileFinder::FindDefault(dir, item.second))});
+	const auto* entries = tree.ListDirectory();
+	if (entries) {
+		for (const auto &item: *entries) {
+			if (item.second.type != DirectoryTree::FileType::Regular) {
+				continue;
+			}
+
+			if (item.first.length() == RtPrefix.length() + 3 && lcf::ToStringView(item.first).starts_with(RtPrefix)) {
```
Comment (fmatthew5876) [inline src/fileext_guesser.cpp:45]: `lcf::StringView` -> `StringView`

File: src/filefinder_rtp.cpp [hunk 14]
```diff
@@ -0,0 +1,308 @@
...
+		search_paths.push_back(std::move(tree));
+
+		if (hit_info.empty()) {
+			Output::Debug("The folder does not contain a known RTP!");
+		}
+
+		// Only consider the best RTP hits (usually 100% if properly installed)
+		float best = 0.0;
+		for (const auto& hit : hit_info) {
+			float rate = (float)hit.hits / hit.max;
```
Comment (fmatthew5876) [inline src/filefinder_rtp.cpp:180]: `static_cast<float>(hit.hits) / hit.max` to make the precedence clear

File: src/directory_tree.cpp (hunk 4 above)
Comment (Ghabry) [reply src/directory_tree.cpp:117]: As this was copied over from FileFinder the code is likely 3 or more years old.
Your solution should work and looks better, thanks.
(The directory type in dirent is an extension and some filesystems reports Unknown here, will also add a comment)

File: src/directory_tree.h (hunk 12 above)
Comment (Ghabry) [reply src/directory_tree.h:250]: added the other ops

**Review by fdelapena — APPROVED** [review]

**Comment by Ghabry** [comment]
Because this is a big change lets keep this open for 7 more days until 6th of Feb and when there is no further feedback: Merge.

Will already use the code here as a base for the final filesystem step :)

**Review by fmatthew5876 — APPROVED** [review]

**Comment by rueter37** [comment]
I'm late for giving feedback but I noticed two oddities after this got merged into master: The compiler (gcc 9.3) complains over `#include <memory>` missing in `directory_tree.h` and RTP files are not getting loaded if certain game folders are missing. You can reproduce this in "TestGame-2000" and "TestGame-2003". In "TestGame-2000" the "Backdrop" and "Monster" folders are missing which makes the background and the monster sprites not being loaded in battle even if the RTP is installed. "TestGame-2003" is even more extreme: Because of every folder missing there, basically no assets are being loaded. But this is probably a rare edge case because most games have every asset folder present (even if they are RTP-reliant).

**Comment by Ghabry** [comment]
Thanks for the reports. I will hotfix them next week.

**Comment by rueter37** [comment]
One more oddity: Savegames are not put into the game folder if you save a game. The save file is put into the same directory from where you have started EasyRPG (you can reproduce it using the command line).

**Comment by Ghabry** [comment]
I didn't notice that bug but this is fixed in PR 3 #2455

**Comment by rueter37** [comment]
Good to know, thanks! :+1:

**Comment by Ghabry** [comment]
Well this shows that it was a good idea to do the DirectoryTree before the VFS. One can find bugs early and the VFS uses the Tree.

The RTP one is dumb: Is caused by this line (abort when dir is missing)
https://github.com/EasyRPG/Player/blob/master/src/directory_tree.cpp#L191-L192

I must restructure the ifs so it falls through to the rtp handler.

**Pull request merged**